go/pkg/yolostore/gorm.go

```diff
@@ -0,0 +1,414 @@
+package yolostore
+
+import (
+	"crypto/sha256"
+	"fmt"
+	"strings"
+
+	"berty.tech/yolo/v2/go/pkg/yolopb"
+	"github.com/jinzhu/gorm"
+	"github.com/mr-tron/base58"
+	"go.uber.org/zap"
+	"moul.io/zapgorm"
+)
+
+type Store interface {
+
+	// artifacts store
+	GetArtifactAndBuildByID(id string) (*yolopb.Artifact, error)
+	GetArtifactByID(id string) (*yolopb.Artifact, error)
+	GetAllArtifactsWithoutBundleID() ([]yolopb.Artifact, error)
+	SaveArtifact(artifact *yolopb.Artifact) error
+
+	// build store
+	GetBuildListFilters() (*BuildListFilters, error)
+	GetLastBuild() (*yolopb.Build, error)
+	GetBuildList(bl BuildList) ([]*yolopb.Build, error)
+
+	// batch store
+	GetBatchWithPreloading() (*yolopb.Batch, error)
+	GetBatch() (*yolopb.Batch, error)
+
+	// download store
+	GetDevDumpObjectDownloads() ([]*yolopb.Download, error)
+	CreateDownload(download *yolopb.Download) error
+	GetArtifactDownload(artifactMap map[string]int64) (map[string]int64, error)
+
+	// internal
+	DB() *gorm.DB
+}
+
+type store struct {
+	db *gorm.DB
+}
+
+func NewStore(db *gorm.DB, logger *zap.Logger) (Store, error) {
+	db, err := initDB(db, logger)
+	if err != nil {
+		return nil, err
+	}
+
+	return &store{
+		db: db,
+	}, nil
+}
+
+func (s *store) DB() *gorm.DB { return s.db }
+
+func initDB(db *gorm.DB, logger *zap.Logger) (*gorm.DB, error) {
+	db.SetLogger(zapgorm.New(logger.Named("gorm")))
+	db.Callback().Create().Remove("gorm:update_time_stamp")
+	db.Callback().Update().Remove("gorm:update_time_stamp")
+	db.Callback().Create().Before("gorm:create").Register("yolo_before_create", beforeCreate)
+	db = db.Set("gorm:auto_preload", false)
+	db = db.Set("gorm:association_autoupdate", false)
+	db.BlockGlobalUpdate(true)
+	db.SingularTable(true)
+	db.LogMode(true)
+	if err := db.AutoMigrate(yolopb.AllModels()...).Error; err != nil {
+		return nil, err
+	}
+	return db, nil
+}
+
+func beforeCreate(scope *gorm.Scope) {
+	if !scope.HasColumn("yolo_id") {
+		return
+	}
+
+	field, found := scope.FieldByName("id")
+	if !found {
+		return
+	}
+
+	table := scope.TableName()
+	id := field.Field.String()
+	hash := sha256.Sum256([]byte(id))
+	encoded := base58.Encode(hash[:])
+	prefix := strings.ToLower(table[:1])
+	yoloID := fmt.Sprintf("%s:%s", prefix, encoded)
+	err := scope.SetColumn("yolo_id", yoloID)
+	if err != nil {
+		panic(err)
+	}
+}
+
+func (s *store) GetArtifactAndBuildByID(id string) (*yolopb.Artifact, error) {
+	var artifact *yolopb.Artifact
+	err := s.db.
+		Preload("HasBuild").
+		Preload("HasBuild.HasProject").
+		Preload("HasBuild.HasProject.HasOwner").
+		First(&artifact, "ID = ?", id).
+		Error
+	if err != nil {
+		return nil, fmt.Errorf("store: GetArtifactByID :%w", err)
+	}
+
+	return artifact, nil
+}
+
+func (s *store) GetArtifactByID(id string) (*yolopb.Artifact, error) {
+	var artifact *yolopb.Artifact
+	err := s.db.First(&artifact, "ID = ?", id).Error
+	if err != nil {
+		return nil, fmt.Errorf("store: GetArtifactByID :%w", err)
+	}
+
+	return artifact, nil
```

Comment:
```suggestion
	var artifact yolopb.Artifact
	err := s.db.First(&artifact, "ID = ?", id).Error
	if err != nil {
		return nil, fmt.Errorf("store: GetArtifactByID :%w", err)
	}

	return &artifact, nil
```

Reply:
Be extremely careful, you made the same mistake as before with:
1. non-inited artifact
2. double-pointer in `First(**yolopb.Artifact)`

This code panics